[[ IsDeletable ]] Ensure objects aren't executing script before delete

montegoulding opened the pull request.

Commit:
[[ Bug 17180 ]] Ensure objects aren't executing script before delete

runrevmark commented on engine/src/stack.cpp:
```diff
@@ -1299,17 +1299,70 @@ void MCStack::applyrect(const MCRectangle &nrect)
+    if (t_deletable)
+        t_deletable = !this -> isediting();
+    
+    if (t_deletable && cards != nil)
```
This will cause all objects to be checked for delete twice - remember that top-level objects are owned by the stack, so you only need to check the card object itself for deletability.

montegoulding replied:
Ah so just check the card script depth in stacks then... Ok

runrevmark commented on engine/src/card.cpp:
```diff
@@ -1088,30 +1088,43 @@ void MCCard::timer(MCNameRef mptr, MCParameter *params)
+{
+    bool t_deletable = true;
+    
+    if (t_deletable)
```
I'd suggest structuring this code as follows:

```
  if (!MCObject::isdeletable())
    return false;
if (!getflag(F_C_CANT_DELETE)
  return false;
...
```

i.e. return false immediately when one of things which causes an object not to be deletable returns false

It makes the code clearer and is also more friendly towards branch prediction (if you have an if with no else, it is generally assumed to be unlikely - and not being able to delete an object is unlikely).

montegoulding replied:
Ok

montegoulding commented on engine/src/card.cpp:
```diff
@@ -1088,30 +1088,43 @@ void MCCard::timer(MCNameRef mptr, MCParameter *params)
+        t_deletable = MCObject::isdeletable();
+    
+    if (t_deletable)
+        t_deletable = !getflag(F_C_CANT_DELETE);
```
I'm thinking this may very well be a significant change of behavior we might or might not want to make where you can't delete a stack if it contains a group, card or substack (or any of the substack's groups and cards) with cantDelete true.

This was previously unchecked but I think it conforms to the docs:

> If an object's cantDelete property is set to true, the object cannot be deleted either by user action or by a handler. If you want to delete the object, you must first set its cantDelete to false.

I'm wondering if/how we can identify the specific object causing the error in this case as it's not necessarily in the trace like scriptdepth identifies. Ideally an error message would say "cantDelete of object <long name> true". I also need to check how this stuff works with revert and anything the IDE does to ensure a revert works as a result of user actions like "Close and remove from memory"... like setting the cantDelete of the mainstack but not substacks, cards or groups...

montegoulding replied:
Hmm... the revert command has no documented interaction with cantDelete however it would have been failing silently previously on stacks with cantDelete true and the other previously checked conditions in `del()`:

```
Boolean oldlock = MClockmessages;
    MClockmessages = True;
    MCerrorlock++;
    if (p_stack->del())
        p_stack -> scheduledelete();
    MCerrorlock--;
    MClockmessages = old lock;
```

Actually I'm wondering if revert silently failing is one of the reasons for _stack name is already loaded hell_ syndrome because it continues after the silent failure of `del()` to load the stack from filename.

```
p_stack = MCdispatcher->findstackname(*t_name);
    if (p_stack != NULL)
        p_stack->openrect(oldrect, oldmode, NULL, WP_DEFAULT, OP_NONE);
```

I'm not really sure why we are locking errors here.

montegoulding replied:
**In the IDE**
- **revFrontScript** 
  - `revDeleteStacks` should ensure all `cantDelete` properties are false and possibly use send in time to ensure the card doesn't have a script depth
- **revMenubar**
  - `revMenubarContextMenu` presents a contextual menu offering the following without checking the cantDelete:
    - remove mainstack from memory
    - delete substack
    - delete card
    - delete group
- **revBackscriptLibrary**
  - `reloadStack` (should be in front script?) set all `cantDelete` properties to false and use send in time to clear the script depth from the stack sent the `reloadStack` message. There also appears to be some unnecessary code duplication in there also so sort that out. _Stack name is already loaded hell_ syndrome is a horrible seemingly impossible to get out of recursive loop that people often complain about. I think I can see why it happens here.
  - `__ideRemoveFromMemory` check the `cantDelete` and deal gracefully with errors
  - `revIDEDeleteObjects`  check the `cantDelete` and deal gracefully with errors
  - `revIDEDeleteCard`   check the `cantDelete` and deal gracefully with errors
  - `revCheckGroupDelete` probably should check the `cantDelete` of the group and put up dialog
  - `revIDEUpdatePlugins` should ensure all `cantDelete` properties are false
  - `revIDEMoveCard`should ensure all `cantDelete` properties are false and reset after to original settings and possibly use send in time to ensure the card doesn't have a script depth.
  - `revIDEMoveControl`should ensure all `cantDelete` properties are false and reset after to original settings and possibly use send in time to ensure the card doesn't have a script depth. Also there's two implementations so we should delete one!!!!
  - `revIDEDeleteObjects` should ensure all `cantDelete` properties are false and reset after to original settings and possibly use send in time to ensure the card doesn't have a script depth
- **revSaveAsStandalone**
  - `revStandaloneDeployWithParams` should ensure all `cantDelete` properties are false and possibly use send in time to ensure the card doesn't have a script depth
  - `revCopyExternals` should ensure all `cantDelete` properties are false and possibly use send in time to ensure the card doesn't have a script depth
  - `revCleanBuildStacks` should ensure all `cantDelete` properties are false and possibly use send in time to ensure the card doesn't have a script depth
  - `revSaveStackFile` should ensure all `cantDelete` properties are false and possibly use send in time to ensure the card doesn't have a script depth
  - `revCloseOpenStacks` should ensure all `cantDelete` properties are false and possibly use send in time to ensure the card doesn't have a script depth
  - `revSBResaveScriptOnlyStackAsProperStackFile` should ensure all `cantDelete` properties are false and possibly use send in time to ensure the card doesn't have a script depth
- **Script Editor** - I got an error from the script editor and I suspect it is because something is being deleted script depth with a non-0 script depth. It needs investigation but here is where getting an error with the actual object causing the problem would be very helpful!

**In the engine:**
- The `revert` command behavior with `this stack` from within the scripts of the stack is completely broken by this PR (I think quite reasonably!). Previously it may have worked if the command was called from a script of the stack other than the stack script. However, this seems very unsafe to me and I think we should kill it! With fire 🔥 

For example:

```
-- button script
on mouseUp
   revert
end mouseUp
```
- Perhaps we need a variation on `delete <object>`... `delete <object> with force`? Would ignore cantDelete flags but not script depth and other dangerous things. This would remove some code from the IDE and avoid having to add a recursive handler that ensured an object didn't have any cantDelete props that would block delete.

montegoulding replied:
Think about this probably deleting a stack should be allowed if the cantDelete of a card or group of the stack is true as deleting from memory <> deleting from stack. I think this would probably avoid too many headaches from this change and avoid needing a force delete option. I'll need to look into how to do it but I suspect the cantDelete of a background shouldn't stop a card it is on from being deleted either.

runrevmark replied:
Yes - you should be able to delete a stack regardless of the cantDelete status of its children. As you point out, deleting a stack is a little different from deleting an object on a stack - the former is removal from memory, the latter is obliteration.

To be fair, I think that there is a confusion in the current engine between 'user action commands' and 'script commands'. Why should script not be allowed to delete an object with cantDelete true? It makes no sense - script should be able to do whatever it wants.

The 'delete object' script command is trying to be both the 'user action' (which should be affected by cantDelete - which is more an editing property) and the general script command.

montegoulding replied:
I think you may be thinking of cantModify rather than cantDelete.

livecodefraser commented:
@livecode-vulcan review okay 36cb29f

livecode-vulcan commented:
💙 review by @livecodefraser ok 36cb29f9514548a569a1ae24bb38cc3036f0dd07

livecode-vulcan commented:
😎 test success 36cb29f9514548a569a1ae24bb38cc3036f0dd07
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2638)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2648)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2633)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2649)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2655)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2661)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2673)

montegoulding commented:
@livecodefraser I'm not sure why the test failed on travis. 

@livecode-vulcan review not ok 36cb29f

livecode-vulcan commented:
💔 review by @montegoulding not ok 36cb29f9514548a569a1ae24bb38cc3036f0dd07

montegoulding commented:
Well I'm not sure why the test is OK here and fails on the server. This test failing means that the object calling a handler that then deletes the object doesn't cause an error to be thrown. I really need more output from travis. Was there a different error number thrown or no error?

montegoulding commented:
OK, it's interesting as in the IDE my test errors correctly but when run against a standalone it crashes.

montegoulding commented:
Aha! It is crashing during `TestTeardown` because I'm deleting the defaultStack and quit is trying to message it. I suppose deleting a stack should mess with `MCdefaultstackptr`. I'll investigate but it appears unrelated to the bug I was fixing.

montegoulding commented:
I added checks to the stack del and destructor for whether it was the topStack because that would have ended up setting `MCdefaultstackptr` back to `this` again but that didn't actually fix it... The fix was `!MCdefaultstackptr->getstate(CS_DELETE_STACK)` before getting the card to message it in quit. However, I can't really see how it is left in a state where it points to a deleted stack given I've also added the getstate check when determining the top stack...

livecodefraser commented:
@livecode-vulcan review okay d8700b7

livecode-vulcan commented:
💙 review by @livecodefraser ok d8700b70da2a556ae572683d5375a7f566e253aa

livecode-vulcan commented:
😎 test success d8700b70da2a556ae572683d5375a7f566e253aa
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/2648)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/2658)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/2643)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/2659)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/2665)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/2671)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/2683)
